### ISSUE TITLE
Images don't show in dev mode with Laravel backend

### ISSUE THREAD

**Issue by HassanZahirnia** [JonasKruckenberg/imagetools#396]
If you try this plugin with a Laravel + Vite + Vue stack, all the images will throw a `404 (Not Found)` error and not show on the page.

This is because the tool is returning relative paths and normally an asset URL imported through Vite would have an absolute URL, and in the context of a backend such as Laravel the lack of absolute URL breaks the images.

Assuming you're running in `dev` mode
Without `imagetools` loader ( normal image importing ):
```ts
import Cat from '@/images/cat.png'

console.log(Cat)
// Output: http://localhost:3000/resources/images/cat.png
```
With `imagetools` loader:
```ts
import Cat from '@/images/cat.png?width=200'

console.log(Cat)
// Output: /@imagetools/5812d5ee51f96cb9c239acca9f2ecc22bd83998a
// Throws a not found error.
```

Note that `build` works fine. It's only a `dev` mode issue.

**Comment by benmccann** [comment]
Can you check if https://github.com/JonasKruckenberg/imagetools/pull/400 would fix the issue you're facing?

**Comment by HassanZahirnia** [comment]
> Can you check if #400 would fix the issue you're facing?

No unfortunately it does not fix the issue :(

Also I think that PR is for a different purpose as they stated:

> This PR fixes the paths of images in the Vite build result.

They talking about the **build** result. This issue I'm having happens in dev mode.

**Comment by hopkins385** [comment]
Same here. Getting 404 Errors on all images when running yarn dev in laravel vite environment


Error:
```shell
Failed to load resource: the server responded with a status of 404 (Not Found)
```
html code that corresponds to error
```html
<source type="image/avif" srcset="/@imagetools/23ae5086f5b8fa6391e57bb0800aabc686bd2b37 850w">
```

### Environment

Console Output: (just for reference and version numbers)
```shell
  VITE v4.0.1  ready in 503 ms

  ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
  ➜  press h to show help

  LARAVEL v9.43.0  plugin v0.7.1
```

package.json
```json
  "laravel-vite-plugin": "^0.7.1",
  "vite": "^4.0.1",
  "vite-imagetools": "^4.0.11",
  "vue": "^3.2.45"
```

vite.config.js
```js
export default defineConfig({
    build: {
        chunkSizeWarningLimit: 600,
    },
    plugins: [
        laravel({
            input: ['resources/css/frontend.css', 'resources/js/frontend.ts'],
            refresh: true,
            buildDirectory: '/build/frontend',
            config: 'frontend',
        }),
        vue({
            template: {
                transformAssetUrls: {
                    base: null,
                    includeAbsolute: false,
                },
            },
        }),
        imagetools({
            /* options */
        }),
    ],
    resolve: {
        alias: {
            '@': path.resolve(__dirname, './resources/js'),
        },
    },
    server: {
        strictPort: true,
        hmr: {
            host: 'localhost',
        },
    },
```

**Comment by DeedleFake** [comment]
I have the same exact problem with a Rails backend:

![image](https://user-images.githubusercontent.com/326750/212525769-02f6b297-f3f4-4937-8be3-5569ab0e971b.png)
![image](https://user-images.githubusercontent.com/326750/212525799-4b56b2ac-a5fe-4ea5-8e8e-5297a3a79534.png)

**Comment by DeedleFake** [comment]
@HassanZahirnia, if you get a chance, could you double-check that the changes from the above-referenced pull request also fixed it for you with Laravel, please?

**Comment by hopkins385** [comment]
nope, unfortunately it does not work. I'm using this plugin in a Laravel 9 and Nuxt 3 environment, in both environments @imagetools is not resolved as url in dev but build works as expected.

**Comment by HassanZahirnia** [comment]
@DeedleFake No the issue is not fixed. Honestly not sure why you marked this issue as "fixed" before checking with a Laravel backend first 😅

If possible please re-open this issue @benmccann

**Comment by HassanZahirnia** [comment]
Hey I've found a similar Vite plugin that uses Sharp underhood, and it's asset paths are correct in dev mode. I thought maybe someone could have a look at it and figure out what they did that would fix this issue, here is the source file:
https://github.com/jack-weilage/vite-image/blob/main/src/index.ts

Although I should mention I could have their plugin work in production. So be careful not to grab any production logic from their code.
Either way the focus is on how they've handled the path for dev mode.

**Comment by DeedleFake** [comment]
> Honestly not sure why you marked this issue as "fixed" before checking with a Laravel backend first 😅

To be honest, I'm not sure why I did either. Just wasn't thinking. Sorry.

**Comment by benmccann** [comment]
@DeedleFake did you test https://github.com/JonasKruckenberg/imagetools/pull/489 with any app? I didn't think about this before, but I'm pretty sure https://github.com/JonasKruckenberg/imagetools/pull/489/files#diff-8d1a92b43a2575de2cdad4edc1af1e9586a5f1ee84ec2dad9b7af5ebc2322b13R123 is wrong because Vite's base middleware strips out the base path before passing it to any middlewares. I'm wondering if https://github.com/JonasKruckenberg/imagetools/pull/489 helps at all or should be reverted entirely?

**Comment by DeedleFake** [comment]
> did you test #489 with any app?

Yes, I did. I tested it with the Rails app that I was having the problems I posted above in this issue and it fixed them.

**Comment by DeedleFake** [comment]
> Vite's base middleware strips out the base path before passing it to any middlewares

The issue with the Rails backend was that the Vite dev server never even saw the requests because they didn't start with the base path. The way the Rails integration works is something like this:

* If the path starts with the Vite base path, send it to the dev server.
* If it starts with something else, check for configured Rails routes and use one that matches.
* If none of those match, check for a matching file in the `public` directory and serve it.
* If all of the above fails, yield a 404.

So without prefixing the string paths returned by the plugin when it's used in an `import` in the code, the Vite dev server will never even see them in the first place.

**Comment by DeedleFake** [comment]
I've never worked with Laravel before, but I set up a simple project and it looks like the problem there is different, but related. In Laravel's case, assets are not routed through the main dev server but are instead sent directly to Vite. For example, in my case the Laravel dev server is running on port 8000, while the Vite dev server is running on port 5173. However, there's a big discrepancy between normal Vite image imports and ones routed through `vite-imagetools`:

```javascript
import test from './test.png'
// test is http://[::1]:5173/resources/js/test.png

import test from './test.png?width=50'
// test is @imagetools/0078fa8134eed9d2fc44c1cd1ab4f62b63fe1041
```

I would speculate that there's another config option for Vite that Laravel is automatically setting to redirect assets to the dev server but that is being missed by `vite-imagetools`. I don't see anything obvious in the config, though.

**Comment by HassanZahirnia** [comment]
@DeedleFake Thanks for looking into Laravel 🙌
The situation you mentioned is same as the one I also described in the first post. The assets need to have an absolute path, except on dev server the path start with the Vite server url (which would be `http://localhost:3000` or `http://[::1]:5173`)

I suggest you to look into the repo I linked earlier. That plugin has a working path setup for dev server. I personally have no experience with Vite plugins so I couldn't figure it out myself.

**Comment by DeedleFake** [comment]
> I suggest you to look into the repo I linked earlier. That plugin has a working path setup for dev server.

I've been doing so, as well as looking at other plugins. The documentation around this particular situation, despite how common you'd think it is, isn't fantastic, unfortunately.

> I personally have no experience with Vite plugins so I couldn't figure it out myself.

Neither do I. I'd never looked into how they worked at all before I started working on fixing this. In fact, I'd never used Vite before a week ago or so.

**Comment by DeedleFake** [comment]
I tried out the repo that you linked above and it has the exact same problem:

![image](https://user-images.githubusercontent.com/326750/213782576-ef0640e0-c457-44b5-8935-867f189f6b4e.png)
![image](https://user-images.githubusercontent.com/326750/213782667-4ef8d202-0546-45bd-96ba-7c35aa1c787f.png)

Neither plugin generates absolute paths for their on-demand assets, and I'm not sure what the proper way to do so is.

**Comment by benmccann** [comment]
> didn't think about this before, but I'm pretty sure https://github.com/JonasKruckenberg/imagetools/pull/489/files#diff-8d1a92b43a2575de2cdad4edc1af1e9586a5f1ee84ec2dad9b7af5ebc2322b13R123 is wrong because Vite's base middleware strips out the base path before passing it to any middlewares

Okay, I think your change was correct and necessary. I can confirm that images in SvelteKit were broken on the prior release when using a base path. The way the imagetools middleware is being called it will be invoked before Vite's base middleware strips out the the base path. If on the otherhand we were returning a function that returned the middleware then we'd no longer want to strip out the base path because Vite would do it: https://vitejs.dev/guide/api-plugin.html#configureserver

**Comment by benmccann** [comment]
Are you using `laravel-vite-plugin` 0.7.3? I see [the latest release](https://github.com/laravel/vite-plugin/releases/tag/v0.7.3) has a fix for the base path, so it might be broken if you're on an older version

**Comment by HassanZahirnia** [comment]
> I tried out the repo that you linked above and it has the exact same problem:
That's weird, it worked for me when I tried it.


> Are you using `laravel-vite-plugin` 0.7.3?
Yeah I'm using the latest. But I have not modified base config so that PR fix does not affect me. Not to mention I believe majority of Laravel users don't change the base config since the default is good enough.


Honestly I think until one of maintainers sit down and try this library with a Laravel backend, there is no way to debug and fix this issue properly.

**Comment by benmccann** [comment]
Could you provide a repository that reproduces the issue?

**Comment by HassanZahirnia** [comment]
@benmccann Here is a reproduction repo: https://github.com/HassanZahirnia/laravel-imagetools-issue

Just in case you're not familiar with Laravel projects here are the instructions:

1. Rename `.env.example` to `.env`
2. Run `composer install`
3. Run `npm install`
4. Run `php artisan key:generate`
5. Run `php artisan serve` in a terminal and keep it open
6. In another terminal run `npm run dev` and also keep it open
7. Visit the url given by the `serve` command (step 5)
8. You can see the image fails to load since it's going through `vite-imagetools` plugin via the `?webp`
9. The file you want to edit/play with is `resources/views/pages/welcome.vue`

**Comment by hopkins385** [comment]
After updating to vite-imagetools v4.0.16 I can confirm the images are getting resolved in Dev mode \o/, in a Nuxt 3 environment. 👍  (Haven't checked Laravel yet)

**Comment by HassanZahirnia** [comment]
Glad the tool is starting working with other frameworks such as Nuxt and Rails, but unfortunately the Laravel support is not there yet.

@hopkins385 I checked with `4.0.16` and it still doesn't work.

**Comment by hopkins385** [comment]
Agree, I've checked it with Laravel and still not working.

package.json (just for reference)

```json
 "vite": "^4.0.4",
 "laravel-vite-plugin": "^0.7.3",
 "vite-imagetools": "^4.0.16",
 "imagetools-core": "^3.3.0",

```

**Comment by benmccann** [comment]
@driesvints @timacdonald any chance you'd be interested in taking a look at this one? It seems to occur exclusively with `laravel-vite-plugin`. For context, `vite-imagetools` is a Vite plugin for optimizing images. A Laravel-based reproduction was provided just above (https://github.com/JonasKruckenberg/imagetools/issues/396#issuecomment-1399288052), but I don't have PHP or Composer installed on my machine. I'm happy to help from this side.

**Comment by timacdonald** [comment]
Hey folks, I have a working solution for this for the Laravel plugin.

For the Laravel folks in the thread, we will can shift out of here and into this issue: https://github.com/laravel/vite-plugin/issues/194

tl;dr; We need to expose a way for developers to say "Please prefix `/@imagetools` with the Vite dev server URL while in dev mode.

```diff
- <img src="/@imagetools/f0b2f404b13f052c604e632f2fb60381bf61a520">
+ <img src="http://[::1]:5173/@imagetools/f0b2f404b13f052c604e632f2fb60381bf61a520">
```

PR will be up shortly.

**Comment by timacdonald** [comment]
See https://github.com/laravel/vite-plugin/pull/195
